Fix corner_pixels set from QtViewer.on_draw when layers have mixed numbers of dimensions

grlee77 opened the pull request:
# Description

This PR is related to #2487.  I don't think it fully closes the issue, but it fixes the problem pointed out by @VolkerH https://github.com/napari/napari/issues/2487#issuecomment-932076245. 

The source of the issue was that `QtViewer`'s `on_draw` method was using the wrong axes when sending the corner pixels to the Layer's `_update_draw` method. 

A separate bug was related to the shape of `corner_pixels` for non-multiscale layers. In `__init__` it is shape `(2, ndim)`:
https://github.com/napari/napari/blob/2f6c4377a3435ac68ed79b9a2e7c7bd0f1730070/napari/layers/base/base.py#L281
, but for non-multiscale images this was getting truncated to shape `(2, 2)` in `_update_draw`:
https://github.com/napari/napari/blob/2f6c4377a3435ac68ed79b9a2e7c7bd0f1730070/napari/layers/base/base.py#L1389

The change here preserves the original `(2, ndim)` shape, making the multiscale and non-multiscale cases consistent.


## Type of change
<!-- Please delete options that are not relevant. -->
- [x] Bug-fix (non-breaking change which fixes an issue)


# References
<!-- What resources, documentation, and guides were used in the creation of this PR? -->
<!-- If this is a bug-fix or otherwise resolves an issue, reference it here with "closes #(issue)" -->

# How has this been tested?
<!-- Please describe the tests that you ran to verify your changes. -->
- [ ] example: the test suite for my feature covers cases x, y, and z
- [ ] example: all tests pass with my change

I checked that the example case given in https://github.com/napari/napari/issues/2487#issuecomment-932076245 works after this change. 

This still needs a unit test, but I could use some help on how best to write it.  I initially tried writing a unit test using a Viewer created via the `make_napari_viewer` fixture. This did not replicate the issue, though, because when running the test there is no `QtViewer` who's `on_draw` method gets called. 


## Final checklist:
- [x] My PR is the minimum possible work for the desired functionality
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] If I included new strings, I have used `trans.` to make them localizable.
      For more information see our [translations guide](https://napari.org/docs/dev/guides/translations.html).

grlee77 committed:
fix displayed_axes offset into _canvas_corners_in_world

When layer.ndim < self.viewer.dims.ndim, have to slice the *last* layer.ndim
dimensions of _canvas_corners_in_world

grlee77 committed:
preserve shape of self.corner_pixels when calling _update_draw

grlee77 committed:
fix dtype of corner_pixels

sofroniewn reviewed (approved):
LGTM! Is there a test you think we could add for this?

grlee77 committed:
Fix calculation of displayed_axes for the general case

The previous setting was not robust against different settings for viewer.dims.order
_canvas_corners_in_world has the same number of dimensions as the viewer, so we want
to select based on the last len(layer._displayed_axes) dimensions.

grlee77 committed:
fix case wher len(layer._displayed_axes) > viewer.dims.ndisplay

need this to allow change back from 3D to 2D view

grlee77 committed:
properly account for dimension order in Layer._update_draw

grlee77 committed:
TST: add test case for corner pixels with mixed 2D and 3D data

grlee77 committed:
Mark test_roll_transpose_update test case as sync_only

This should not be necessary longer term, and should be revisited after future
async refactoring.
